linear_operator/operators/block_diag_linear_operator.py

```diff
@@ -153,8 +153,9 @@ def matmul(self, other):
             return BlockDiagLinearOperator(self.base_linear_op @ other.base_linear_op)
         # special case if we have a DiagLinearOperator
         if isinstance(other, DiagLinearOperator):
-            diag_reshape = other._diag.view(*self.base_linear_op.shape[:-2], 1, -1)
-            return BlockDiagLinearOperator(self.base_linear_op * diag_reshape)
+            diag_reshape = other._diag.view(*self.base_linear_op.shape[:-1])
+            diag = DiagLinearOperator(diag_reshape)
+            return BlockDiagLinearOperator(self.base_linear_op @ diag)
```

Comment:
may be worth adding a comment here that this matmul is going to be cheap b/c of the special casing in `DiagLinearOperator`?